src/main/kotlin/gg/flyte/christmas/minigame/games/Spleef.kt

```diff
@@ -10,155 +11,214 @@ import gg.flyte.christmas.util.eventController
+
+    private var doubleJumps = mutableMapOf<UUID, Int>()
+
+    private var unlimitedJumps = false
+    private var remainingUnlimitedJumpTicks = 0
+
+    private val unlimitedJumpBar = BossBar.bossBar("<gold><b>UNLIMITED DOUBLE JUMPS".style(), 1.0F, BossBar.Color.BLUE, BossBar.Overlay.NOTCHED_6)
+    private var unlimitedJumpBarTicks = 0.0F
+
+    private var powerfulSnowballs = false
+    private var remainingPowerfulSnowballTicks = 0
+
+    private val snowballBar = BossBar.bossBar("<gold><b>POWERFUL SNOWBALLS".style(), 1.0F, BossBar.Color.WHITE, BossBar.Overlay.NOTCHED_6)
+    private var snowballBarTicks = 0.0F
+
+    private val snowmen = mutableListOf<Snowman>()
+    private val bees = mutableListOf<Bee>()
+
+    private var bottomLayerMelted = false
 
     override fun startGameOverview() {
         super.startGameOverview()
 
         overviewTasks += repeatingTask(1) {
             while (floorLevelBlocks.random().block.type != Material.AIR) {
-                floorLevelBlocks.random().block.type = Material.AIR
+                wearDownSnowBlock(floorLevelBlocks.random().block, true)
             }
         }
     }
 
     override fun preparePlayer(player: Player) {
         player.formatInventory()
-        player.gameMode = GameMode.ADVENTURE
-        player.teleport(gameConfig.spawnPoints.random().randomLocation())
+
         ItemStack(Material.DIAMOND_SHOVEL).apply {
             itemMeta = itemMeta.apply {
                 displayName("<!i><game_colour>Snow Shovel!".style())
             }
         }.let { player.inventory.setItem(0, it) }
-        doubleJumps[player] = 3
+
+        player.gameMode = GameMode.ADVENTURE
+        player.teleport(gameConfig.spawnPoints.random().randomLocation())
+
+        doubleJumps[player.uniqueId] = 3
     }
 
     override fun startGame() {
         overviewTasks.forEach { it.cancel() }
+
         donationEventsEnabled = true
 
-        for (point in floorLevelBlocks) point.block.type = Material.SNOW_BLOCK // reset after game overview
+        for (point in floorLevelBlocks) {
+            point.block.type = Material.SNOW_BLOCK
+        } // reset after game overview
 
         simpleCountdown {
             Util.runAction(PlayerType.PARTICIPANT) {
                 it.gameMode = GameMode.SURVIVAL
                 it.allowFlight = true
             }
 
+            //prevent snow golems from forming snow layers
+            ChristmasEventPlugin.instance.serverWorld.setGameRule(GameRule.MOB_GRIEFING, false)
+
+            addActionBarTask()
+            addUnlimitedJumpsTask()
+            addPowerfulSnowballsTask()
+            addSnowmenSetTargetTask()
+
+            tasks += repeatingTask(1) {
+                remainingPlayers().forEach {
+                    if (it.location.blockY < 70) {
+                        if (remainingPlayers().contains(it)) {
+                            eliminate(it, EliminationReason.ELIMINATED)
+                        }
+                    }
+                }
```

Comment:
Runnables are probably more efficient than `PlayerMoveEvent`, also since I need to check for bees and snowmen too, I'd have to use `EntityMoveEvent` to keep consistency which is even worse.